### PR TITLE
MemoryPoolMXBean implementation

### PR DESCRIPTION
**Pull request by petermz**
Hi --
This is a fix for https://github.com/oracle/graal/issues/4510. It introduces 2 memory pool beans representing young and old heap generations. Usage numbers are provided for current usage, peak usage, and collection usage.

In `MemoryUsage` objects returned by this implementation, it is always the case that `used == committed`. This is not exactly true as some memory is usually taken up by unused chunks, which makes it committed but unused. However, those unused chunks are shared between young and old generations, and cannot be attributed to either pool.

Typical output of the test program from the above issue now looks like this:
```
found 2 gc
==> gc young generation scavenger CollectionCount=21 CollectionTime=955 MemoryPoolNames=[young generation space]
==> gc complete scavenger CollectionCount=9 CollectionTime=938 MemoryPoolNames=[young generation space, old generation space]
found 2 memory pools
==> memoryPoolMXBean young generation space type=Heap memory isValid=true getMemoryManagerNames=[young generation scavenger, complete scavenger] usage=init = 0(0K) used = 90177536(88064K) committed = 90177536(88064K) max = 357564416(349184K) peak usage=init = 0(0K) used = 179306496(175104K) committed = 179306496(175104K) max = 357564416(349184K) collection usage=init = 0(0K) used = 0(0K) committed = 0(0K) max = 357564416(349184K)
==> memoryPoolMXBean old generation space type=Heap memory isValid=true getMemoryManagerNames=[complete scavenger] usage=init = 0(0K) used = 141539552(138222K) committed = 141539552(138222K) max = 716177408(699392K) peak usage=init = 0(0K) used = 350206176(341998K) committed = 350206176(341998K) max = 716177408(699392K) collection usage=init = 0(0K) used = 141539552(138222K) committed = 141539552(138222K) max = 716177408(699392K)
```

### CODE REVIEW

**Comment by petermz** [comment]
Fixed style issues -- the checks should now pass

**Comment by igoragueme** [comment]
some news about this issue/pr? @christianhaeubl

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/graal/GenScavengeGCFeature.java [hunk 1]
```diff
@@ -92,6 +94,7 @@ public void duringSetup(DuringSetupAccess access) {
 
         ManagementSupport managementSupport = ManagementSupport.getSingleton();
         managementSupport.addPlatformManagedObjectSingleton(java.lang.management.MemoryMXBean.class, new HeapImplMemoryMXBean());
+        managementSupport.addPlatformManagedObjectList(java.lang.management.MemoryPoolMXBean.class, Arrays.asList(new YoungGenerationMemoryPoolMXBean(), new OldGenerationMemoryPoolMXBean()));
```
Comment (christianhaeubl) [inline substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/graal/GenScavengeGCFeature.java:97]: HotSpot uses three pools for the serial GC: `eden`, `survivor`, and `old`. I think we should do the same.

`GenScavengeGCFeature` is used for the serial and for the epsilon GC. For the epsilon GC, only a single memory pool should be added (like on HotSpot).

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGenerationMemoryPoolMXBean.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import org.graalvm.nativeimage.Platforms;
+
+/**
+ * A MemoryPoolMXBean for the young generation.
+ */
+public final class YoungGenerationMemoryPoolMXBean extends AbstractMemoryPoolMXBean {
+
+    @Platforms(Platform.HOSTED_ONLY.class)
+    public YoungGenerationMemoryPoolMXBean() {
+        super("young generation space", "young generation scavenger", "complete scavenger");
```
Comment (christianhaeubl) [inline substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGenerationMemoryPoolMXBean.java:40]: The manager names should be passed as an argument into the constructor of `YoungGenerationMemoryPoolMXBean` and `OldGenerationMemoryPoolMXBean`. This avoids the duplication of the string `complete scavenger` and makes sure that all the manager names are visible in one place.

**Review by christianhaeubl — COMMENTED** [review]
Thanks for working on that. I added a few comments.

File: substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/AbstractMemoryPoolMXBean.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+import sun.management.Util;
+
+public abstract class AbstractMemoryPoolMXBean extends MXBeanBase implements MemoryPoolMXBean {
+
+    protected final String name;
+    protected final String[] managerNames;
+    protected final String objectName;
+    protected final GCAccounting gcAccounting;
+
+    protected AbstractMemoryPoolMXBean(String name, String... managerNames) {
```
[the remark this answers is not in the document]
Comment (petermz) [reply substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/AbstractMemoryPoolMXBean.java:44]: Unfortunately I need this class to send notifications from `GCAccounting`. I can of course define a dedicated interface, but is it worth doing?

**Comment by petermz** [comment]
Thanks Christian -- I've implemented most of the changes

**Review by christianhaeubl — APPROVED** [review]
Thanks. I did a few small cleanups and fixes on top. Next week, your PR will be merged to master.

**Pull request merged**